### PR TITLE
Fix helicopter rotor collides at twice the rotor diameter

### LINKED ISSUE #39474: Helicopter rotor collides at twice the rotor diameter
# Describe the bug
Helicopter rotor collision code appears to calculate collisions at twice the actual rotor diameter. This appears to be caused by using the diameter as radius: https://github.com/CleverRaven/Cataclysm-DDA/blob/f594632cd6a5d9dac884c3807eebb2357d42548d/src/vehicle_move.cpp#L421

# Steps To Reproduce
Came across this while flying for the first time and attempting to land a medevac helicopter at the military helipad map location. This helicopter has "small helicopter rotors" with a diameter of 8 according to the current rotor.json file. If I land with any object within (roughly) the purple circle the helicopter will crash. See attached screenshot. 

This location has a two story building which I tested approaching at the 2nd floor. Getting any closer than 8(+1) squares from the rotor center caused the helicopter to crash with the wall. Note how the purple circle extends past the tailboom of the helicopter which appears to be a rare design choice in real life due to interfering with the (imaginary) tail rotor. As such I consider this a bug.

# Expected behavior
Expect diameter/radius mismatch to be corrected and helicopter rotor diameter to fit within radius of tailboom on default heli designs. Would be helpful with an indicator of some kind to tell where the rotor is as I'm certain the player character has at least some idea from their position in the craft.

Another related issue: low objects like fuelpumps appear to cause rotor collisions which may not be intended.

# Screenshots
helicopter_rotor_diameter

# Versions and configuration
 - OS: Windows 7
 - Game Version: 0.E-468-g9e02bc1
 - Graphics version: Tiles
 - Ingame language: English
 - Mods loaded: dda, no_npc_food, aftershock, crt_expansion, magiclysm, alt_map_key, No_Fungi, fuji_mpp, StatsThroughSkills

PS: great job on the vehicle crashes, they're really quite spectacular. :)

### PR DESCRIPTION
**Pull request by olanti-p**
#### Summary
SUMMARY: Bugfixes "Fix helicopter rotor collides at twice the rotor diameter"

#### Purpose of change
Fixes #39474

#### Describe the solution
Divide by 2

#### Testing
Spawned a helicopter with rotor diameter 8 and placed a wall 5 tiles away.
heli
Ascended, then descended. The helicopter lands safely with this fix but crashes without it.

### CODE REVIEW

**Comment by kevingranade** [comment]
Round up please, integer division implicitly floors.

**Comment by olanti-p** [comment]
It should, shouldn't it? If D=9, you count 4.5 tiles from the center of the rotor and have the same situation as in the pic

File: src/vehicle_move.cpp [hunk 1]
```diff
@@ -419,7 +419,8 @@ bool vehicle::collision( std::vector<veh_collision> &colls,
         const tripoint dsp = global_pos3() + dp + parts[p].precalc[1];
         veh_collision coll = part_collision( p, dsp, just_detect, bash_floor );
         if( coll.type == veh_coll_nothing && info.rotor_diameter() > 0 ) {
-            for( const tripoint &rotor_point : g->m.points_in_radius( dsp, info.rotor_diameter() ) ) {
+            size_t radius = static_cast<size_t>( info.rotor_diameter() / 2 );
```
Comment (ZhilkinSerg) [inline src/vehicle_move.cpp:422]:
```suggestion
            size_t radius = static_cast<size_t>( std::round( info.rotor_diameter() / 2.0f ) );
```

**Comment by olanti-p** [comment]
Sorry for a lack of response, I was sleeping.
Flooring was not a division error, it was intentional.

Let's say `D = 2K` or `D = 2K + 1`. Then `K = floor(D/2.0)`. `g->m.points_in_radius(R)` returns an iterator over a `(2R+1)x(2R+1)` area. If we pass it `K`, we get an iterator over `DxD` for `D = 2K + 1` and `(D+1)x(D+1)` for `D = 2K`. Example values:
D=1 ..... area=1x1
D=2 ..... area=3x3
D=3 ..... area=3x3
D=4 ..... area=5x5
D=5 ..... area=5x5

Let's say `D = 2K + 1`. We round it up: `L = round(D/2.0) = K + 1`. Then we pass `L` to the function and receive an iterator over `(2*(K+1)+1)x(2*(K+1)+1)`, i.e. `(D+2)x(D+2)`. Example values:
D=1 ..... area=3x3
D=2 ..... area=3x3
D=3 ..... area=5x5
D=4 ..... area=5x5
D=5 ..... area=7x7

For a rotor with D=5 to have a collision area of 7x7 would be unexpected for the players.

**Comment by ZhilkinSerg** [comment]
Let's change `rotor diameter` to `blade length` maybe? It would be `g->m.points_in_radius( dsp, blade_length )` (i.e. `blade_length * 2 + 1`) then.

File: src/vehicle_move.cpp [hunk 2]
```diff
@@ -419,7 +419,8 @@ bool vehicle::collision( std::vector<veh_collision> &colls,
         const tripoint dsp = global_pos3() + dp + parts[p].precalc[1];
         veh_collision coll = part_collision( p, dsp, just_detect, bash_floor );
         if( coll.type == veh_coll_nothing && info.rotor_diameter() > 0 ) {
-            for( const tripoint &rotor_point : g->m.points_in_radius( dsp, info.rotor_diameter() ) ) {
+            size_t radius = static_cast<size_t>( std::round( info.rotor_diameter() / 2.0f ) );
```
Comment (jbytheway) [inline src/vehicle_move.cpp:422]: Suggest using `divide_round_up` rather than floating-point arithmetic here.

**Comment by olanti-p** [comment]
Sounds good to me. The lift thrust calculation can be easily adapted to use `blade length` too. I can make it a PR

**Comment by kevingranade** [comment]
> Flooring was not a division error, it was intentional.

I understand, I just disagree.
When rotor diameter exceeds 1m (it always will), it necessarally enters tiles adjacent to it's origin.  In order for that to happen when D = 1.1, R = f( D ) must be greater than 1.
From there, as D increases, R obviously increases at half the rate of D.
Given these invariants, the most straightforward definition of R is ceil( D / 2 ).

From a balance point of view, anything we do here is incredibly generous, from what I've read about helipad layout, minimum safe landing zone footprint is 2D.   It is very likely that we will add up to several tiles of horizontal jitter when ascending and descending to represent this.  Because of this, I'm not particularly concerned about the, "players will be surprised that they can't land in a tiny area" issue, because that is a fundamentally unsafe thing to do.  If you want to alleviate the surprise factor there, we can look into emitting a, "minimum safe distance" number, but that number should be significantly larger than the physical rotor footprint.

**Comment by olanti-p** [comment]
So it basically adds 0.5-1 tile to real radius as a penalty for tiny jitter? Can't argue that. Should be mentioned somewhere in the code probably.
```
If you want to alleviate the surprise factor there
```
No, it's fine. It was more of a "this code produces wrong R for given D since helicopters are absolutely still" issue. When (if) the jitter is added, the surprise factor will no longer be there since helicopters would be visibly affected by jitter/wind.

**Pull request merged**